homeassistant/components/madvr/config_flow.py

```diff
@@ -36,81 +32,120 @@ class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
+    ) -> ConfigFlowResult:
+        """Handle reconfiguration of the device."""
+        self.entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        return await self.async_step_reconfigure_confirm(user_input)
+
+    async def async_step_reconfigure_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration flow initialized by the user."""
+        return await self._handle_config_step(user_input, reconfigure=True)
+
+    async def _handle_config_step(
+        self, user_input: dict[str, Any] | None = None, reconfigure: bool = False
+    ) -> ConfigFlowResult:
+        """Handle the configuration step for both initial setup and reconfiguration."""
         errors: dict[str, str] = {}
+        step_id = "reconfigure" if reconfigure else "user"
 
         if user_input is not None:
+            _LOGGER.debug("User input: %s", user_input)
             host = user_input[CONF_HOST]
             port = user_input[CONF_PORT]
 
             try:
-                # ensure we can connect and get the mac address from device
-                mac = await self._test_connection(host, port)
+                mac = await test_connection(self.hass, host, port)
             except CannotConnect:
                 _LOGGER.error("CannotConnect error caught")
                 errors["base"] = "cannot_connect"
             else:
                 if not mac:
                     errors["base"] = "no_mac"
-            if not errors:
-                _LOGGER.debug("MAC address found: %s", mac)
-                # this will prevent the user from adding the same device twice and persist the mac address
-                await self.async_set_unique_id(mac)
-                self._abort_if_unique_id_configured()
-
-                # create the entry
-                return self.async_create_entry(
-                    title=DEFAULT_NAME,
-                    data=user_input,
-                )
-
-        # this will show the form or allow the user to retry if there was an error
+                else:
+                    _LOGGER.debug("MAC address found: %s", mac)
+                    # if mac does change, it should be updated
+                    await self.async_set_unique_id(mac)
+
+                    if reconfigure:
+                        _LOGGER.debug("Reconfiguring device")
+                        if TYPE_CHECKING:
+                            assert self.entry is not None
+                        self.hass.config_entries.async_update_entry(
+                            self.entry,
+                            data={**user_input, CONF_HOST: host, CONF_PORT: port},
+                        )
+                        await self.hass.config_entries.async_reload(self.entry.entry_id)
+                        _LOGGER.debug("Reconfiguration successful")
+                        return self.async_abort(reason="reconfigure_successful")
```

Comment:
There's a helper for this, don't know the name but I think `discovergy` uses it

Reply:
yep 

async_update_reload_and_abort